examples/server_data_source.rs

```diff
@@ -0,0 +1,118 @@
+use std::{
+    sync::{Arc, Mutex},
+    thread,
+};
+
+use anyhow::Context as _;
+use open62541::{ua, DataSource, ObjectNode, Server, VariableNode};
+use open62541_sys::{
+    UA_NS0ID_BASEDATAVARIABLETYPE, UA_NS0ID_FOLDERTYPE, UA_NS0ID_OBJECTSFOLDER, UA_NS0ID_ORGANIZES,
+    UA_NS0ID_STRING,
+};
+
+fn main() -> anyhow::Result<()> {
+    env_logger::init();
+
+    let (server, runner) = Server::new();
+
+    println!("Adding server nodes");
+
+    let object_node = ObjectNode {
+        requested_new_node_id: ua::NodeId::string(1, "the.folder"),
+        parent_node_id: ua::NodeId::ns0(UA_NS0ID_OBJECTSFOLDER),
+        reference_type_id: ua::NodeId::ns0(UA_NS0ID_ORGANIZES),
+        browse_name: ua::QualifiedName::new(1, "the folder"),
+        type_definition: ua::NodeId::ns0(UA_NS0ID_FOLDERTYPE),
+        attributes: ua::ObjectAttributes::default(),
+    };
+
+    let variable_node_id = ua::NodeId::string(1, "the.answer");
+    let variable_node = VariableNode {
+        requested_new_node_id: variable_node_id.clone(),
+        parent_node_id: object_node.requested_new_node_id.clone(),
+        reference_type_id: ua::NodeId::ns0(UA_NS0ID_ORGANIZES),
+        browse_name: ua::QualifiedName::new(1, "the answer"),
+        type_definition: ua::NodeId::ns0(UA_NS0ID_BASEDATAVARIABLETYPE),
+        attributes: ua::VariableAttributes::default()
+            .with_data_type(&ua::NodeId::ns0(UA_NS0ID_STRING))
+            .with_access_level(
+                &ua::AccessLevel::NONE
+                    .with_current_read(true)
+                    .with_current_write(true),
+            ),
+    };
+
+    let current_value = Arc::new(Mutex::new(String::from("Lorem ipsum")));
+
+    let data_source = DataSource::read_write(
```

Comment:
Probably yes.

A `trait` object would make the shared state explicit instead of implicitly capturing it. The `context` then becomes `self`.

Not sure about other implications and restrictions.